README.md
# nagios-cookbook

TODO: Enter the cookbook description here.

## Supported Platforms

TODO: List your supported platforms.

## Attributes

<table>
  <tr>
    <th>Key</th>
    <th>Type</th>
    <th>Description</th>
    <th>Default</th>
  </tr>
  <tr>
    <td><tt>['nagios']['bacon']</tt></td>
    <td>Boolean</td>
    <td>whether to include bacon</td>
    <td><tt>true</tt></td>
  </tr>
</table>

## Usage

### nagios::default

Include `nagios` in your node's `run_list`:

```json
{
  "run_list": [
    "recipe[nagios::default]"
  ]
}
```

## License and Authors

Author:: Wanelo, Inc. (<dev@wanelo.com>)


CHANGELOG.md
# 0.1.0

Initial release of nagios


test/cookbooks/test-setup/metadata.rb
name 'test-setup'
maintainer 'test'
maintainer_email 'test'
license 'MIT'
description 'cookbook for setting up tests'
long_description 'cookbook for setting up tests'
version '0.0.1'

depends 'nagios'


test/cookbooks/test-setup/recipes/install.rb
include_recipe 'nagios::default'

nagios 'nagios' do
  server_name 'nagios.example.com'
  action :install
end

execute 'proxy server name in hosts file' do
  command 'echo "127.0.0.1 nagios.example.com" >> /etc/hosts'
  not_if 'grep nagios.example.com /etc/hosts'
end


test/integration/install/serverspec/service_spec.rb
require 'spec_helper'

RSpec.describe 'nagios::install' do
  describe service('nagios') do
    it { should be_running }
  end
end


recipes/default.rb
#
# Cookbook Name:: nagios
# Recipe:: default
#
# Copyright (C) 2015 Wanelo, Inc.
#
# Permission is hereby granted, free of charge, to any person obtaining
# a copy of this software and associated documentation files (the
# "Software"), to deal in the Software without restriction, including
# without limitation the rights to use, copy, modify, merge, publish,
# distribute, sublicense, and/or sell copies of the Software, and to
# permit persons to whom the Software is furnished to do so, subject to
# the following conditions:
#
# The above copyright notice and this permission notice shall be
# included in all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
# LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
# OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
# WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
#


recipes/service.rb
include_recipe 'smf'

execute 'verify nagios' do
  action :nothing
  command 'nagios -v /opt/local/etc/nagios/nagios.cfg'
  environment 'PATH' => node['paths']['bin_path']
  notifies :restart, 'service[nagios]'
end

template '/opt/local/sbin/nagios-dependencies' do
  mode 0755
end

smf 'pkgsrc/nagios' do
  action :delete
end

smf 'nagios-dependencies' do
  fmri '/application/setup/nagios-dependencies'
  start_command '/opt/local/sbin/nagios-dependencies'
  stop_command 'true'
  duration 'transient'
  manifest_type 'setup'
  dependencies [
      {'name' => 'multi-user', 'fmris' => ['svc:/milestone/multi-user'],
        'grouping' => 'require_all', 'restart_on' => 'none', 'type' => 'service'}
    ]
  notifies :run, 'execute[verify nagios]'
  notifies :enable, 'service[nagios-dependencies]', :immediately
end

smf 'nagios' do
  fmri '/application/nagios'
  user 'nagios'
  group 'nagios'
  start_command '/opt/local/bin/nagios -d %{config_file}'
  start_timeout 60
  stop_timeout 60
  ignore %w(core signal)
  working_directory '/var/spool/nagios'
  property_groups({
      'application' => {
        'config_file' => '/opt/local/etc/nagios/nagios.cfg'
      }
    })

  dependencies [
      {'name' => 'nagios-dependencies', 'fmris' => ['svc:/application/setup/nagios-dependencies'],
        'grouping' => 'require_all', 'restart_on' => 'none', 'type' => 'service'}
    ]
  notifies :run, 'execute[verify nagios]'
end

service 'nagios'
service 'nagios-dependencies'


libraries/nagios_provider.rb
require 'chef/mixin/shell_out'

class Chef
  class Provider
    # Provider for the nagios Chef provider
    #
    # nagios 'nagios' do
    #   service_name 'nagios.example.com'
    # end
    #
    class Nagios < Chef::Provider::LWRPBase
      include Chef::Mixin::ShellOut

      def load_current_resource
        @current_resource ||= new_resource.class.new(new_resource.name)
      end

      def action_install
        new_resource.updated_by_last_action(false)

        install_packages
        configure_nagios
        install_service
        configure_nginx
      end

      def action_restart
        Chef::Log.warn "Nagios: Restarting service"
        new_resource.notifies_immediately(:restart, nagios_service)
        new_resource.updated_by_last_action(true)
      end

      def action_verify
        Chef::Log.warn "Nagios: Verifying configuration"

        execute 'verify nagios configuration' do
          command '/opt/local/bin/nagios -v /opt/local/etc/nagios/nagios.cfg'
          environment 'PATH' => node['paths']['bin_path']
          notifies :restart, new_resource.to_s, :immediately
        end
        new_resource.updated_by_last_action(true)
      end

      private

      def configure_nagios
        template '/opt/local/etc/nagios/cgi.cfg' do
          source 'nagios/cgi.cfg.erb'
          cookbook 'nagios'
          owner 'nagios'
          group 'nagios'
          mode 0644
          notifies :verify, new_resource.to_s
        end
      end

      def configure_nginx
        run_context.include_recipe 'nginx'

        template '/opt/local/etc/nginx/sites-available/nagios.conf' do
          source 'nginx/nagios.conf.erb'
          cookbook 'nagios'
          owner 'www'
          group 'www'
          mode 0644
          variables 'server_name' => new_resource.server_name
          notifies :run, 'execute[verify nginx configuration file]'
        end

        nginx_site 'nagios.conf' do
          enable true
        end

        execute 'verify nginx configuration file' do
          command 'nginx -t -c /opt/local/etc/nginx/nginx.conf'
          action :nothing
          notifies :reload, 'service[nginx]'
        end
      end

      def install_packages
        %w(php55-fpm spawn-fcgi fcgiwrap nagios-base).each do |pkg|
          package pkg
        end
      end

      def install_service
        run_context.include_recipe 'smf'

        smf 'pkgsrc/nagios' do
          action :delete
        end

        template '/opt/local/sbin/nagios-dependencies' do
          source 'nagios-dependencies.erb'
          cookbook 'nagios'
          mode 0755
        end

        smf 'nagios-dependencies' do
          fmri '/application/setup/nagios-dependencies'
          start_command '/opt/local/sbin/nagios-dependencies'
          stop_command 'true'
          duration 'transient'
          manifest_type 'setup'
          dependencies [
              {'name' => 'multi-user', 'fmris' => ['svc:/milestone/multi-user'],
                'grouping' => 'require_all', 'restart_on' => 'none', 'type' => 'service'}
            ]
          notifies :enable, 'service[nagios-dependencies]', :immediately
          notifies :verify, new_resource.to_s
        end

        service 'nagios-dependencies'

        smf 'nagios' do
          fmri '/application/nagios'
          user 'nagios'
          group 'nagios'
          start_command '/opt/local/bin/nagios -d %{config_file}'
          start_timeout 70
          stop_timeout 60
          ignore %w(core signal)
          working_directory '/var/spool/nagios'
          property_groups({
              'application' => {
                'config_file' => '/opt/local/etc/nagios/nagios.cfg'
              }
            })

          dependencies [
              {'name' => 'nagios-dependencies', 'fmris' => ['svc:/application/setup/nagios-dependencies'],
                'grouping' => 'require_all', 'restart_on' => 'none', 'type' => 'service'}
            ]
          notifies :verify, new_resource.to_s
        end
      end

      def nagios_service
        begin
          run_context.resource_collection.find(:service => 'nagios')
        rescue Chef::Exceptions::ResourceNotFound
          service 'nagios' do
            supports restart: true, status: true, reload: true
          end
        end
      end
    end
  end
end
